hypervisor/util.go

```diff
@@ -0,0 +1,69 @@
+package hypervisor
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+)
+
+type Volume struct {
+	Path    string
+	Format  string // raw|qcow2|...
+	AioType string // native|threads
+	Cache   string // none|unsafe|writethrough...
+}
+
+// parseVolumes parses --volume strings that are of following format:
+// --volume {volumePath}[:{options}]
+// Example: --volume /path/to/myvolume.img:format=raw:aio=native
+func ParseVolumes(volumeStrings []string) ([]Volume, error) {
+	res := []Volume{}
+	if volumeStrings == nil {
+		return res, nil
+	}
+	for _, volumeStr := range volumeStrings {
+		if v, err := parseVolume(volumeStr); err == nil {
+			res = append(res, *v)
+		} else {
+			return res, err
+		}
+
+	}
+	return res, nil
+}
+
+func parseVolume(volumeStr string) (*Volume, error) {
+	v := &Volume{
+		Path:    "",
+		Format:  "raw",
+		AioType: "native",
+		Cache:   "none",
+	}
+
+	for idx, part := range strings.Split(volumeStr, ":") {
+		if idx == 0 { // Volume path
+			if path, err := filepath.Abs(part); err == nil {
+				v.Path = path
+				continue
+			} else {
+				return nil, err
+			}
+		} else { // Volume settings
+			if !strings.Contains(volumeStr, "=") {
+				return nil, fmt.Errorf("missing '=' in volume setting")
```

Comment:
I suspect this gets output to the user. If that is so, i suggest printing out something in the lines of "Please use '=' for assignment of volume settings. Example: .... ". If this does not get produced as output to the user ignore my comment.

Reply:
On second thought, you are doing this check each time in the loop. You do not need to, as you are doubling work. Do it before the loop, or check each part (not the whole volumeStr) in the loop.

Reply:
Sharp eye! The `part` was to be checked indeed, not the `volumeStr`. Also thanks for suggestion regarding the error message. I remember a year ago or so how with @gberginc we also did another review iteration just to provide a better message. And ever since we're reading the message many times and we're always glad to have it rewritten 😄